File: server/src/main/java/au/com/codeka/warworlds/server/admin/handlers/EmpireDetailsHandler.java
package au.com.codeka.warworlds.server.admin.handlers;

import com.google.common.collect.ImmutableMap;

import java.util.ArrayList;
import java.util.List;

import au.com.codeka.warworlds.common.proto.DeviceInfo;
import au.com.codeka.warworlds.common.proto.Empire;
import au.com.codeka.warworlds.common.proto.Notification;
import au.com.codeka.warworlds.common.proto.Star;
import au.com.codeka.warworlds.server.handlers.RequestException;
import au.com.codeka.warworlds.server.proto.PatreonInfo;
import au.com.codeka.warworlds.server.store.DataStore;
import au.com.codeka.warworlds.server.world.NotificationManager;
import au.com.codeka.warworlds.server.world.StarManager;
import au.com.codeka.warworlds.server.world.WatchableObject;

/**
 * Handler for /admin/empires/xxx which shows details about the empire with id xxx.
 */
public class EmpireDetailsHandler extends AdminHandler {
  @Override
  public void get() throws RequestException {
    long id = Long.parseLong(getUrlParameter("id"));
    Empire empire = DataStore.i.empires().get(id);
    if (empire == null) {
      throw new RequestException(404);
    }

    complete(empire, ImmutableMap.builder());
  }

  @Override
  public void post() throws RequestException {
    long id = Long.parseLong(getUrlParameter("id"));
    Empire empire = DataStore.i.empires().get(id);
    if (empire == null) {
      throw new RequestException(404);
    }

    String msg = getRequest().getParameter("msg");
    if (msg.length() == 0) {
      complete(empire, ImmutableMap.<String, Object>builder()
          .put("error", "You need to specify a message."));
      return;
    }

    // TODO: send it
    NotificationManager.i.sendNotification(empire, new Notification.Builder()
        .debug_message(msg)
        .build());

    redirect("/admin/empires/" + id);
  }

  private void complete(Empire empire, ImmutableMap.Builder<String, Object> mapBuilder)
      throws RequestException {
    mapBuilder.put("empire", empire);

    ArrayList<Star> stars = new ArrayList<>();
    for (Long starId : DataStore.i.stars().getStarsForEmpire(empire.id)) {
      WatchableObject<Star> star = StarManager.i.getStar(starId);
      if (star != null) {
        stars.add(star.get());
      }
    }
    mapBuilder.put("stars", stars);

    mapBuilder.put("devices", DataStore.i.empires().getDevicesForEmpire(empire.id));

    PatreonInfo patreonInfo = DataStore.i.empires().getPatreonInfo(empire.id);
    if (patreonInfo != null) {
      mapBuilder.put("patreon", patreonInfo);
    }

    render("empires/details.html", mapBuilder.build());
  }
}

File: server/src/main/kotlin/au/com/codeka/warworlds/server/admin/handlers/EmpireDetailsHandler.kt
package au.com.codeka.warworlds.server.admin.handlers

import au.com.codeka.warworlds.common.proto.*
import au.com.codeka.warworlds.common.sim.DesignDefinitions
import au.com.codeka.warworlds.server.handlers.RequestException
import au.com.codeka.warworlds.server.proto.PatreonInfo
import au.com.codeka.warworlds.server.store.DataStore
import au.com.codeka.warworlds.server.world.NotificationManager
import au.com.codeka.warworlds.server.world.StarManager
import au.com.codeka.warworlds.server.world.WatchableObject
import java.util.*

/**
 * Handler for /admin/empires/xxx which shows details about the empire with id xxx.
 */
class EmpireDetailsHandler : AdminHandler() {
  public override fun get() {
    val id = getUrlParameter("id")!!.toLong()
    val empire: Empire = DataStore.i.empires()[id] ?: throw RequestException(404)

    val data = HashMap<String, Any>()
    data["empire"] = empire

    when (request.getParameter("tab")) {
      "stars" -> {
        completeStarsTab(empire, data)
      }
      "devices" -> {
        completeDevicesTab(empire, data)
      }
      "sit-reports" -> {
        completeSitReportsTab(empire, data)
      }
      else -> {
        completeOverview(empire, data)
      }
    }
  }

  public override fun post() {
    val id = getUrlParameter("id")!!.toLong()
    val empire: Empire = DataStore.i.empires()[id] ?: throw RequestException(404)
    val msg = request.getParameter("msg")
    if (msg.isEmpty()) {
      val data = HashMap<String, Any>()
      data["empire"] = empire
      data["error"] = "You need to specify a message."
      completeOverview(empire, data)
      return
    }

    // TODO: send it
    NotificationManager.i.sendNotification(empire, Notification(debug_message = msg))
    redirect("/admin/empires/$id")
  }

  private fun completeOverview(empire: Empire, data: HashMap<String, Any>) {
    data["account"] = DataStore.i.accounts().getByEmpireId(empire.id)!!.two

    val patreonInfo: PatreonInfo? = DataStore.i.empires().getPatreonInfo(empire.id)
    if (patreonInfo != null) {
      data["patreon"] = patreonInfo
    }

    render("empires/details.html", data)
  }

  private fun completeStarsTab(empire: Empire, data: HashMap<String, Any>) {
    val stars = ArrayList<Star?>()
    for (starId in DataStore.i.stars().getStarsForEmpire(empire.id)) {
      val star: WatchableObject<Star>? = StarManager.i.getStar(starId)
      if (star != null) {
        stars.add(star.get())
      }
    }
    data["stars"] = stars

    render("empires/details-stars.html", data)
  }

  private fun completeDevicesTab(empire: Empire, data: HashMap<String, Any>) {
    data["devices"] = DataStore.i.empires().getDevicesForEmpire(empire.id)
    render("empires/details-devices.html", data)
  }

  private fun completeSitReportsTab(empire: Empire, data: HashMap<String, Any>) {
    val sitReports = DataStore.i.sitReports().getByEmpireId(empire.id, 50)
    data["sitReports"] = sitReports

    val designs = HashMap<String, Design>()
    for (design in DesignDefinitions.designs.designs) {
      designs[design.type.name] = design
    }
    data["designs"] = designs

    val stars = HashMap<Long, Star>()
    for (sitReport in sitReports) {
      if (stars.containsKey(sitReport.star_id)) {
        continue
      }

      val star = StarManager.i.getStar(sitReport.star_id)
      if (star != null) {
        stars[sitReport.star_id] = star.get()
      }
    }
    data["stars"] = stars

    render("empires/details-sitreports.html", data)
  }
}
